### ISSUE TITLE
Do not warn about mixing a string and an array in some cases

### ISSUE THREAD

**Issue by jzinn** [koalaman/shellcheck#1249]
- SC2145 Argument mixes string and array
- version: 0.5.0
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh

subcommand_foo() {
  echo "In foo"
  echo "\$1 == $1 == aaa"
  echo "\$2 == $2 == bbb"
}

subcommand_bar() {
  echo "In bar"
}

subcommand_baz() {
  echo "In baz"
}

main() {
  case "$1" in
    foo | bar | baz )
      subcommand_"$@"
    ;;
    * )
      echo "Error: Unrecognized command"
    ;;
  esac
}

main foo aaa bbb
```

#### Here's what shellcheck currently says:

```
$ shellcheck myscript
 
Line 20:
      subcommand_"$@"
                  ^-- SC2145: Argument mixes string and array. Use * or separate argument.

$ 
```

#### Here's what I wanted or expected to see:

IMO, this is a legitimate use of mixing a string and an array.  Perhaps a special case can be made for instances like this so that shellcheck does not warn about it.

**Comment by jzinn** [comment]
By the way, I edited the wiki, expecting that GitHub would make a fork and a pull request, but the edit saved successfully.  The page is

*   https://github.com/koalaman/shellcheck/wiki/SC2145

Please revert it if you do not like the content.

**Comment by matthewpersico** [comment]
case "$1" in
      foo | bar | baz )
        func=$1
        shift
        subcommand_"$func" "$@"
    ;;

Problem solved.  More obvious what you are doing.

**Comment by koalaman** [comment]
You're right in that this is a working and intentional mixing of strings and arrays. You've checked that there is at least one argument, and you know how `$@` behaves in the middle of words.

How can ShellCheck differentiate this from the problematic cases like `--var="$@"` or `printf "You wrote $@\\n"`?

**Comment by jzinn** [comment]
Differentiating might be tricky, and I haven't given it serious thought.  In the absence of a clever idea for that, always warning is reasonable.  Using a ShellCheck disable comment for the few cases where it's intentional is not a problem.

One condition to look for might be when the string + first element of the array would expand to a command (i.e. there's no other string before the string in question).  Maybe this test is sufficient, but I'm not certain.  For your `--var="$@"` example, in the wild there would likely be some string before it, such as `foo` in `foo --var="$@"`, that would become the command.

If there is text between the `$@` and the close quote, then that should be warned about.  For instance, warn about `"blah${@}foo" "bar"` because of the foo, but not because of the `bar`.

**Comment by koalaman** [comment]
This seems very case specific, and I don't know how well it would generalize. Let's let the human make the call through directives until we figure out a good heuristic.
